receiver/kafkareceiver/kafka_receiver.go

```diff
@@ -0,0 +1,168 @@
+			return nil, err
+		}
+		c.Version = version
+	}
+	client, err := sarama.NewConsumerGroup(config.Brokers, config.GroupID, c)
+	if err != nil {
+		return nil, err
+	}
+	return &kafkaConsumer{
+		name:          config.Name(),
+		consumerGroup: client,
+		topics:        []string{config.Topic},
+		nextConsumer:  nextConsumer,
+		unmarshaller:  &protoUnmarshaller{},
+		logger:        params.Logger,
+	}, nil
+}
+
+func (c *kafkaConsumer) Start(context.Context, component.Host) error {
+	ctx, cancel := context.WithCancel(context.Background())
+	c.cancel = cancel
+	consumerGroup := &consumerGroupHandler{
+		name:         c.name,
+		logger:       c.logger,
+		unmarshaller: c.unmarshaller,
+		nextConsumer: c.nextConsumer,
+		ready:        make(chan bool),
+	}
+	go c.consumeLoop(ctx, consumerGroup)
+	<-consumerGroup.ready
+	return nil
+}
+
+func (c *kafkaConsumer) consumeLoop(ctx context.Context, handler sarama.ConsumerGroupHandler) error {
+	for {
+		// `Consume` should be called inside an infinite loop, when a
+		// server-side rebalance happens, the consumer session will need to be
+		// recreated to get the new claims
+		if err := c.consumerGroup.Consume(ctx, c.topics, handler); err != nil {
+			c.logger.Error("Error from consumer", zap.Error(err))
+		}
+		// check if context was cancelled, signaling that the consumer should stop
+		if ctx.Err() != nil {
+			c.logger.Info("Consumer stopped", zap.Error(ctx.Err()))
+			return ctx.Err()
+		}
+	}
+}
+
+func (c *kafkaConsumer) Shutdown(context.Context) error {
+	c.cancel()
+	return c.consumerGroup.Close()
+}
+
+type consumerGroupHandler struct {
+	name         string
+	unmarshaller unmarshaller
+	nextConsumer consumer.TraceConsumer
+	ready        chan bool
+
+	logger *zap.Logger
+}
+
+var _ sarama.ConsumerGroupHandler = (*consumerGroupHandler)(nil)
+
+func (c *consumerGroupHandler) Setup(session sarama.ConsumerGroupSession) error {
+	close(c.ready)
+	statsTags := []tag.Mutator{tag.Insert(tagInstanceName, c.name)}
+	_ = stats.RecordWithTags(session.Context(), statsTags, statPartitionStart.M(1))
+	return nil
+}
+
+func (c *consumerGroupHandler) Cleanup(session sarama.ConsumerGroupSession) error {
+	statsTags := []tag.Mutator{tag.Insert(tagInstanceName, c.name)}
+	_ = stats.RecordWithTags(session.Context(), statsTags, statPartitionClose.M(1))
+	return nil
+}
+
+func (c *consumerGroupHandler) ConsumeClaim(session sarama.ConsumerGroupSession, claim sarama.ConsumerGroupClaim) error {
+	c.logger.Info("Starting consumer group", zap.Int32("partition", claim.Partition()))
+	for message := range claim.Messages() {
+		session.MarkMessage(message, "")
+		ctx := obsreport.ReceiverContext(session.Context(), c.name, transport, c.name)
+		ctx = obsreport.StartTraceDataReceiveOp(ctx, c.name, transport)
+		c.logger.Debug("Kafka message claimed",
+			zap.String("value", string(message.Value)),
+			zap.Time("timestamp", message.Timestamp),
+			zap.String("topic", message.Topic))
+
+		// TODO shall we use processor tag?
```

Comment:
No, why processor, this is a receiver....

Reply:
done